File: mac/Sources/SeanboyCore/S3/S3Client.swift
import Foundation

/// Connection details for an S3-compatible bucket. Built for Cloudflare R2
/// (`https://<account>.r2.cloudflarestorage.com`, region `auto`) but works
/// against any S3 host, including MinIO on a NAS.
public struct S3Config: Codable, Equatable, Sendable {
    public var endpoint: URL
    public var bucket: String
    public var region: String
    public var accessKeyID: String
    public var secretAccessKey: String

    public init(endpoint: URL, bucket: String, region: String = "auto",
                accessKeyID: String, secretAccessKey: String) {
        self.endpoint = endpoint
        self.bucket = bucket
        self.region = region
        self.accessKeyID = accessKeyID
        self.secretAccessKey = secretAccessKey
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        endpoint = try container.decode(URL.self, forKey: .endpoint)
        bucket = try container.decode(String.self, forKey: .bucket)
        region = try container.decodeIfPresent(String.self, forKey: .region) ?? "auto"
        accessKeyID = try container.decode(String.self, forKey: .accessKeyID)
        secretAccessKey = try container.decode(String.self, forKey: .secretAccessKey)
    }
}

public struct S3Object: Equatable, Sendable {
    public var key: String
    public var etag: String

    public init(key: String, etag: String) {
        self.key = key
        self.etag = etag
    }
}

public enum S3Error: Error, LocalizedError, Equatable {
    /// A conditional write failed — the object changed under us (or already
    /// exists for If-None-Match). The next sync re-merges and heals this.
    case preconditionFailed(key: String)
    case notFound(key: String)
    case http(status: Int, message: String)
    case badResponse

    public var errorDescription: String? {
        switch self {
        case .preconditionFailed(let key):
            return "\(key) changed on the server — will re-merge on next sync"
        case .notFound(let key):
            return "\(key) does not exist in the bucket"
        case .http(let status, let message):
            return message.isEmpty ? "Server returned HTTP \(status)" : message
        case .badResponse:
            return "Unexpected response from the server"
        }
    }
}

/// Minimal async S3 client: list / get / put / delete / copy, with
/// conditional writes. Paths are bucket-in-path style so a single endpoint
/// URL is all the configuration a provider needs.
public final class S3Client: Sendable {
    public let config: S3Config
    private let session: URLSession

    public init(config: S3Config, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    // MARK: - Operations

    /// Lists every object under `prefix`, following continuation tokens.
    public func list(prefix: String) async throws -> [S3Object] {
        var objects: [S3Object] = []
        var continuation: String?
        repeat {
            var query: [(name: String, value: String)] = [
                ("list-type", "2"),
                ("prefix", prefix),
            ]
            if let continuation {
                query.append(("continuation-token", continuation))
            }
            let (data, _) = try await send(
                method: "GET", key: nil, query: query, body: nil)
            let page = ListResultParser.parse(data)
            objects.append(contentsOf: page.objects)
            continuation = page.nextContinuationToken
        } while continuation != nil
        return objects
    }

    public func get(key: String) async throws -> (data: Data, etag: String) {
        let (data, response) = try await send(method: "GET", key: key, query: [], body: nil)
        return (data, etag(from: response, key: key))
    }

    /// Uploads an object. `ifMatch` makes the write conditional on the
    /// current ETag; `ifNoneMatch` requires the key to not exist yet. Either
    /// failing throws `S3Error.preconditionFailed`.
    @discardableResult
    public func put(key: String, data: Data, contentType: String = "text/markdown",
                    ifMatch: String? = nil, ifNoneMatch: Bool = false) async throws -> String {
        var headers = ["content-type": contentType]
        if let ifMatch { headers["if-match"] = "\"\(ifMatch)\"" }
        if ifNoneMatch { headers["if-none-match"] = "*" }
        let (_, response) = try await send(
            method: "PUT", key: key, query: [], body: data, extraHeaders: headers)
        return etag(from: response, key: key)
    }

    /// Idempotent — deleting a missing key succeeds.
    public func delete(key: String) async throws {
        _ = try await send(method: "DELETE", key: key, query: [], body: nil)
    }

    public func copy(from sourceKey: String, to destinationKey: String) async throws {
        let source = "/" + SigV4.encode(config.bucket + "/" + sourceKey, encodeSlash: false)
        _ = try await send(method: "PUT", key: destinationKey, query: [], body: nil,
                           extraHeaders: ["x-amz-copy-source": source])
    }

    // MARK: - Request plumbing

    private func send(method: String, key: String?,
                      query: [(name: String, value: String)], body: Data?,
                      extraHeaders: [String: String] = [:]) async throws
        -> (Data, HTTPURLResponse) {
        guard let host = config.endpoint.host else { throw S3Error.badResponse }

        var path = "/" + config.bucket
        if let key { path += "/" + key }

        let payloadHash = SigV4.hexHash(body ?? Data())
        var headers = extraHeaders
        headers["host"] = host
        headers["x-amz-content-sha256"] = payloadHash

        let signed = SigV4.sign(
            SigV4.Request(method: method, path: path, queryItems: query,
                          headers: headers, payloadHash: payloadHash, date: Date()),
            credentials: .init(accessKeyID: config.accessKeyID,
                               secretAccessKey: config.secretAccessKey),
            region: config.region)
        headers.merge(signed) { _, new in new }
        headers.removeValue(forKey: "host")  // URLSession sets Host itself

        var urlString = config.endpoint.absoluteString
        if urlString.hasSuffix("/") { urlString.removeLast() }
        urlString += SigV4.encode(path, encodeSlash: false)
        let queryString = SigV4.canonicalQuery(query)
        if !queryString.isEmpty { urlString += "?" + queryString }
        guard let url = URL(string: urlString) else { throw S3Error.badResponse }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw S3Error.badResponse }

        switch http.statusCode {
        case 200..<300:
            return (data, http)
        case 412:
            throw S3Error.preconditionFailed(key: key ?? path)
        case 404 where method == "DELETE":
            return (data, http)  // deleting a missing key is fine
        case 404:
            throw S3Error.notFound(key: key ?? path)
        default:
            throw S3Error.http(status: http.statusCode,
                               message: ErrorResultParser.message(from: data))
        }
    }

    private func etag(from response: HTTPURLResponse, key: String) -> String {
        let raw = response.value(forHTTPHeaderField: "ETag") ?? ""
        return raw.trimmingCharacters(in: CharacterSet(charactersIn: "\""))
    }
}

// MARK: - XML parsing

/// Parses a ListObjectsV2 result page.
final class ListResultParser: NSObject, XMLParserDelegate {
    struct Page {
        var objects: [S3Object] = []
        var nextContinuationToken: String?
    }

    private var page = Page()
    private var currentElement = ""
    private var currentText = ""
    private var inContents = false
    private var key = ""
    private var etag = ""

    static func parse(_ data: Data) -> Page {
        let delegate = ListResultParser()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        parser.parse()
        return delegate.page
    }

    func parser(_ parser: XMLParser, didStartElement name: String,
                namespaceURI: String?, qualifiedName: String?,
                attributes: [String: String] = [:]) {
        currentElement = name
        currentText = ""
        if name == "Contents" {
            inContents = true
            key = ""
            etag = ""
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentText += string
    }

    func parser(_ parser: XMLParser, didEndElement name: String,
                namespaceURI: String?, qualifiedName: String?) {
        let text = currentText.trimmingCharacters(in: .whitespacesAndNewlines)
        switch name {
        case "Key" where inContents:
            key = text
        case "ETag" where inContents:
            etag = text.trimmingCharacters(in: CharacterSet(charactersIn: "\""))
        case "Contents":
            inContents = false
            page.objects.append(S3Object(key: key, etag: etag))
        case "NextContinuationToken":
            page.nextContinuationToken = text
        default:
            break
        }
        currentText = ""
    }
}

/// Pulls the human-readable message out of an S3 error body.
final class ErrorResultParser: NSObject, XMLParserDelegate {
    private var message = ""
    private var code = ""
    private var currentElement = ""
    private var currentText = ""

    static func message(from data: Data) -> String {
        let delegate = ErrorResultParser()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        parser.parse()
        if !delegate.message.isEmpty { return delegate.message }
        return delegate.code
    }

    func parser(_ parser: XMLParser, didStartElement name: String,
                namespaceURI: String?, qualifiedName: String?,
                attributes: [String: String] = [:]) {
        currentElement = name
        currentText = ""
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentText += string
    }

    func parser(_ parser: XMLParser, didEndElement name: String,
                namespaceURI: String?, qualifiedName: String?) {
        let text = currentText.trimmingCharacters(in: .whitespacesAndNewlines)
        if name == "Message" { message = text }
        if name == "Code" { code = text }
        currentText = ""
    }
}

File: android/core/src/main/kotlin/com/torchcodelab/seanboy/core/S3Client.kt
package com.torchcodelab.seanboy.core

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayInputStream
import java.net.URI
import java.time.Instant
import javax.xml.parsers.SAXParserFactory
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler

/**
 * Connection details for an S3-compatible bucket. Built for Cloudflare R2
 * (`https://<account>.r2.cloudflarestorage.com`, region `auto`) but works
 * against any S3 host, including MinIO on a NAS.
 */
data class S3Config(
    val endpoint: String,
    val bucket: String,
    val region: String = "auto",
    val accessKeyID: String,
    val secretAccessKey: String,
)

data class S3Object(val key: String, val etag: String)

/** Result of a GET: the body bytes and the object's ETag. */
data class S3Get(val data: ByteArray, val etag: String)

sealed class S3Error(message: String) : Exception(message) {
    /**
     * A conditional write failed — the object changed under us (or already
     * exists for If-None-Match). The next sync re-merges and heals this.
     */
    data class PreconditionFailed(val key: String) :
        S3Error("$key changed on the server — will re-merge on next sync")

    data class NotFound(val key: String) : S3Error("$key does not exist in the bucket")
    data class Http(val status: Int, val body: String) :
        S3Error(if (body.isEmpty()) "Server returned HTTP $status" else body)

    data object BadResponse : S3Error("Unexpected response from the server")
}

/**
 * Minimal blocking S3 client: list / get / put / delete / copy, with
 * conditional writes. Bucket-in-path style so a single endpoint URL is all the
 * configuration a provider needs. Callers run it off the main thread.
 *
 * Kotlin port of `mac/Sources/SeanboyCore/S3/S3Client.swift` (URLSession/async
 * becomes OkHttp/blocking; Foundation XMLParser becomes SAX).
 */
class S3Client(
    val config: S3Config,
    private val client: OkHttpClient = OkHttpClient(),
) {
    private val host: String =
        URI(config.endpoint).host ?: throw S3Error.BadResponse

    // MARK: - Operations

    /** Lists every object under [prefix], following continuation tokens. */
    fun list(prefix: String): List<S3Object> {
        val objects = mutableListOf<S3Object>()
        var continuation: String? = null
        do {
            val query = mutableListOf("list-type" to "2", "prefix" to prefix)
            continuation?.let { query.add("continuation-token" to it) }
            val (data, _) = send("GET", key = null, query = query, body = null)
            val page = ListResultParser.parse(data)
            objects.addAll(page.objects)
            continuation = page.nextContinuationToken
        } while (continuation != null)
        return objects
    }

    fun get(key: String): S3Get {
        val (data, response) = send("GET", key = key, query = emptyList(), body = null)
        return S3Get(data, etagOf(response))
    }

    /**
     * Uploads an object. [ifMatch] makes the write conditional on the current
     * ETag; [ifNoneMatch] requires the key to not exist yet. Either failing
     * throws [S3Error.PreconditionFailed].
     */
    fun put(
        key: String,
        data: ByteArray,
        contentType: String = "text/markdown",
        ifMatch: String? = null,
        ifNoneMatch: Boolean = false,
    ): String {
        val extra = mutableMapOf("content-type" to contentType)
        if (ifMatch != null) extra["if-match"] = "\"$ifMatch\""
        if (ifNoneMatch) extra["if-none-match"] = "*"
        val (_, response) = send("PUT", key = key, query = emptyList(), body = data, extraHeaders = extra)
        return etagOf(response)
    }

    /** Idempotent — deleting a missing key succeeds. */
    fun delete(key: String) {
        send("DELETE", key = key, query = emptyList(), body = null)
    }

    fun copy(fromKey: String, toKey: String) {
        val source = "/" + SigV4.encode(config.bucket + "/" + fromKey, encodeSlash = false)
        send("PUT", key = toKey, query = emptyList(), body = null, extraHeaders = mapOf("x-amz-copy-source" to source))
    }

    // MARK: - Request plumbing

    private class Resp(val code: Int, val etag: String)

    private fun send(
        method: String,
        key: String?,
        query: List<Pair<String, String>>,
        body: ByteArray?,
        extraHeaders: Map<String, String> = emptyMap(),
    ): Pair<ByteArray, Resp> {
        var path = "/" + config.bucket
        if (key != null) path += "/" + key

        val payloadHash = SigV4.hexHash(body ?: ByteArray(0))
        val headers = extraHeaders.toMutableMap()
        headers["host"] = host
        headers["x-amz-content-sha256"] = payloadHash

        val signed = SigV4.sign(
            SigV4.Request(
                method = method, path = path, queryItems = query,
                headers = headers, payloadHash = payloadHash, date = Instant.now(),
            ),
            credentials = SigV4.Credentials(config.accessKeyID, config.secretAccessKey),
            region = config.region,
        )
        headers.putAll(signed)
        headers.remove("host") // OkHttp sets Host itself

        var urlString = config.endpoint.trimEnd('/')
        urlString += SigV4.encode(path, encodeSlash = false)
        val queryString = SigV4.canonicalQuery(query)
        if (queryString.isNotEmpty()) urlString += "?$queryString"

        val requestBody = when (method) {
            "PUT", "POST" -> (body ?: ByteArray(0)).toRequestBody()
            else -> null
        }
        val builder = Request.Builder().url(urlString).method(method, requestBody)
        for ((name, value) in headers) builder.header(name, value)

        client.newCall(builder.build()).execute().use { response ->
            val bytes = response.body?.bytes() ?: ByteArray(0)
            val resp = Resp(response.code, response.header("ETag").orEmpty().trim('"'))
            return when {
                response.code in 200..299 -> bytes to resp
                response.code == 412 -> throw S3Error.PreconditionFailed(key ?: path)
                response.code == 404 && method == "DELETE" -> bytes to resp
                response.code == 404 -> throw S3Error.NotFound(key ?: path)
                else -> throw S3Error.Http(response.code, ErrorResultParser.message(bytes))
            }
        }
    }

    private fun etagOf(resp: Resp): String = resp.etag
}

// MARK: - XML parsing

/** Parses a ListObjectsV2 result page. */
internal object ListResultParser {
    class Page {
        val objects = mutableListOf<S3Object>()
        var nextContinuationToken: String? = null
    }

    fun parse(data: ByteArray): Page {
        val page = Page()
        val handler = object : DefaultHandler() {
            private val text = StringBuilder()
            private var inContents = false
            private var key = ""
            private var etag = ""

            override fun startElement(uri: String?, localName: String?, qName: String, attrs: Attributes?) {
                text.setLength(0)
                if (qName == "Contents") {
                    inContents = true; key = ""; etag = ""
                }
            }

            override fun characters(ch: CharArray, start: Int, length: Int) {
                text.appendRange(ch, start, start + length)
            }

            override fun endElement(uri: String?, localName: String?, qName: String) {
                val value = text.toString().trim()
                when (qName) {
                    "Key" -> if (inContents) key = value
                    "ETag" -> if (inContents) etag = value.trim('"')
                    "Contents" -> { inContents = false; page.objects.add(S3Object(key, etag)) }
                    "NextContinuationToken" -> page.nextContinuationToken = value
                }
                text.setLength(0)
            }
        }
        runCatching {
            SAXParserFactory.newInstance().newSAXParser()
                .parse(ByteArrayInputStream(data), handler)
        }
        return page
    }
}

/** Pulls the human-readable message out of an S3 error body. */
internal object ErrorResultParser {
    fun message(data: ByteArray): String {
        var message = ""
        var code = ""
        val handler = object : DefaultHandler() {
            private val text = StringBuilder()
            override fun startElement(uri: String?, localName: String?, qName: String, attrs: Attributes?) {
                text.setLength(0)
            }

            override fun characters(ch: CharArray, start: Int, length: Int) {
                text.appendRange(ch, start, start + length)
            }

            override fun endElement(uri: String?, localName: String?, qName: String) {
                val value = text.toString().trim()
                if (qName == "Message") message = value
                if (qName == "Code") code = value
                text.setLength(0)
            }
        }
        runCatching {
            SAXParserFactory.newInstance().newSAXParser()
                .parse(ByteArrayInputStream(data), handler)
        }
        return message.ifEmpty { code }
    }
}
